File: Problems/630_Course_Schedule_III.cpp
#include<vector>
#include<algorithm>
#include<queue>
#include<iostream>

using namespace std;
class Solution {
public:
    static bool compare(vector<int>& m, vector<int>& n){
        return m[1]<n[1];
    }
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), compare);
        priority_queue<int> heap;
        int curr_days = 0;
        for(vector<int> course: courses){
            if(curr_days+course[0] <= course[1]){
                curr_days += course[0];
                heap.push(course[0]);
            }else if(!heap.empty() && heap.top()>course[0]){
                curr_days = curr_days - heap.top() + course[0];
                heap.pop();
                heap.push(course[0]);
            }
        }
        return heap.size();
    }
};

int main(){
    Solution sol;
    vector<vector<int>> courses = {{3,2},{4,3}};
    // cout<<sol.compare(courses)<<endl;
}

File: Problems/630_Course_Schedule_III.py
import heapq
from typing import List

class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        courses.sort(key = lambda x: x[1])
        heap = []
        heapq.heapify(heap)

        curr_days = 0
        for duration, deadline in courses:
            if curr_days + duration <= deadline:
                curr_days += duration
                heapq.heappush(heap, -duration)
            else:
                poped = heapq.heappushpop(heap, -duration)
                curr_days = curr_days + poped + duration
        return len(heap)

if __name__ == '__main__':
    sol = Solution()
    courses = [[1,2],[2,3]]
    print(sol.scheduleCourse(courses))